### PR TITLE
drivers: imx8m: SDMA

### PR DESCRIPTION
**Pull request by paulstelian97**
This patch set introduces the i.MX SDMA driver and enables it on the i.MX8M platform, where it is used.

The driver is mostly clean and complete however there is one issue that has to be accounted for and I'm asking for advice: The hardware reserves channel 0 for management activities and this channel _cannot_ be used for anything else. Should I tell SOF that there are 32 channels but one is occupied? Should I tell that there are only 31? If the latter, should I also shift the channel numbers so the index 0 is actually channel 1 on hardware?

### CODE REVIEW

**Comment by lgirdwood** [comment]
@paulstelian97 Intel uses 1 DMA channel for trace, not sure what logic we are using to reserve/alloc it atm, probably worth doing the same.

**Comment by paulstelian97** [comment]
> @paulstelian97 Intel uses 1 DMA channel for trace, not sure what logic we are using to reserve/alloc it atm, probably worth doing the same.

In that case it's requested from outside. So really is it sufficient what I'm doing here, namely just setting the state so that dma_channel_get() never returns it? I thought the channel counter was used to just skip this controller if all channels are busy.

File: src/platform/imx8m/lib/dma.c [hunk 1]
```diff
@@ -12,6 +12,7 @@
 #include <sof/spinlock.h>
 
 extern struct dma_ops dummy_dma_ops;
+extern struct dma_ops sdma_ops;
```
Comment (dbaluta) [inline src/platform/imx8m/lib/dma.c]: @paulstelian97 can we remove do not merge and add a comment about what needs to be done? In this way we will have booting with qemu ready and we fix the rest later.

Comment (paulstelian97) [reply src/platform/imx8m/lib/dma.c]: Alright. Mentioned in the commit message that a way to account for channel 0 is needed. As for our functional tests not succeeding, we will still need to investigate (perhaps open a GH issue to keep track of it)

Comment (dbaluta) [reply src/platform/imx8m/lib/dma.c]: @paulstelian97 there are still some checkpatch.pl issues, care to have a look?

Comment (paulstelian97) [reply src/platform/imx8m/lib/dma.c]: Resolved most of what I could spot, 2 cannot be solved without significant changes. I need to figure out a way to reduce the indentation in those two cases.

Comment (dbaluta) [reply src/platform/imx8m/lib/dma.c]: @paulstelian97 thanks Paul. Looks good to me now. Will merge this change now after CI does the basic checks.

Comment (paulstelian97) [reply src/platform/imx8m/lib/dma.c]: Fixed more of them as I noticed them in CI.

**Comment by dbaluta** [comment]
@lgirdwood @xiulipan I wonder about CI failed tests. Are they legit? Our PR doesn't touch any intel code.

**Comment by paulstelian97** [comment]
> @lgirdwood @xiulipan I wonder about CI failed tests. Are they legit? Our PR doesn't touch any intel code.

Will look at them when CI is redone, pushed another removal of error traces. Perhaps the error trace from within Probe that I just removed caused our own tests to fail.

We don't touch Intel code so I don't expect any CI errors to be induced there. About our currently-working platforms we may need to test separately.

**Comment by paulstelian97** [comment]
sof-logger errors are due to CI using outdated tool. All others don't seem to be consistent.

**Comment by dbaluta** [comment]
@paulstelian97 yes, noticed that. I think we are safe to merge. Let CI finish required tests once again and I will merge the PR.

**Pull request merged**